File: javascript/rng.js
class RNG {
    constructor(seed) {
        this._a = [0, seed[0], seed[1], seed[2]];
        this._b = [0, seed[3], seed[4], seed[5]];
    }
    get a() {
        return this._a;
    }
    get b() {
        return this._b;
    }
    Next() {
        this._a[0] = (1403580 * this._a[2] - 810728 * this._a[3]) % 4294967087;
        this._b[0] = (527612 * this._b[1] - 1370589 * this._b[3]) % 4294944443;
        var z = (this._a[0] - this._b[0]) % 4294967087;
        if (z < 0) {
            z += 429467087;
        }
        var u = 4294967087.0 / 4294967088.0;
        if (z > 0) {
            u = z / 4294967087.0;
        }
        this._a[3] = this._a[2];
        this._a[2] = this._a[1];
        this._a[1] = this._a[0];
        this._b[3] = this._b[2];
        this._b[2] = this._b[1];
        this._b[1] = this._b[0];
        return u;
    };
}

function main() {
    var seed = [1, 2, 3, 4, 5, 6];
    var gen = new RNG(seed);
    for (var i = 0; i < 10; i++) {
        console.log(gen.Next());
    }
}
main();

File: julia/rng.jl
module RNG
    struct ArenaIter
        a::NTuple{3, Int64}
        b::NTuple{3, Int64}
        function ArenaIter(a::NTuple{3, Int64}, b::NTuple{3, Int64})
            new(a, b)
        end
        # custom constructor
        function ArenaIter(seed::Vararg{Int64, 6})
            new(seed[1:3], seed[4:6])
        end
    end
    function next(iter::ArenaIter)::Tuple{Float64,ArenaIter}
        a = (1403580 * iter.a[2] - 810728 * iter.a[3]) % 4294967087  # might be negative
        b = (527612 * iter.b[1] - 1370589 * iter.b[3]) % 4294944443  # might be negative
        z = (a - b) % 4294967087
        z += (z < 0) ? 4294967087 : 0 # positivify
        u = (z > 0) ? convert(Float64, z) / 4294967088.0 : 4294967087.0 / 4294967088.0
        u, ArenaIter((a, iter.a[1], iter.a[2]), (b, iter.b[1], iter.b[2]))
    end
    # iteration
    function Base.iterate(iter::ArenaIter)::Union{Nothing, Tuple{Float64, ArenaIter}}
        next(iter)
    end
    function Base.iterate(iter::ArenaIter, state::ArenaIter)::Union{Nothing, Tuple{Float64, ArenaIter}}
        next(state)
    end
end